src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,432 @@
+package org.jenkinsci.plugins.pipeline.milestone;
+
+
+import static java.util.logging.Level.WARNING;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.Map.Entry;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.annotation.CheckForNull;
+
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowGraphWalker;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.FlowStartNode;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
+import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+
+import com.google.inject.Inject;
+
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.XmlFile;
+import hudson.model.Executor;
+import hudson.model.InvisibleAction;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.model.Jenkins;
+
+public class MilestoneStepExecution extends AbstractSynchronousStepExecution<Void> {
+
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+
+    @Inject(optional=true) private transient MilestoneStep step;
+    @StepContextParameter private transient Run<?,?> run;
+    @StepContextParameter private transient FlowNode node;
+    @StepContextParameter private transient TaskListener listener;
+
+    @Override
+    public Void run() throws Exception {
+        if (step.getLabel() != null) {
+            node.addAction(new LabelAction(step.getLabel()));
```

Comment:
The `label` still feels superfluous to me, but if you want to keep it, I would suggest also adding `StageAction`. That will cause it to be picked up by https://github.com/jenkinsci/workflow-api-plugin/pull/1 in what I think is the way you would expect.